storage/innobase/trx/trx0i_s.cc

```diff
@@ -819,112 +785,18 @@ static
+	const lock_t&	lock,	/*!< in: lock object to fold */
+	uint16_t	heap_no)/*!< in: lock's record number
 				or 0xFFFF if the lock
 				is a table lock */
 {
-#ifdef TEST_LOCK_FOLD_ALWAYS_DIFFERENT
-	static ulint	fold = 0;
-
-	return(fold++);
-#else
-	ulint	ret;
-
-	if (!lock->is_table()) {
-		ut_a(heap_no != 0xFFFF);
-		ret = ut_fold_ulint_pair((ulint) lock->trx->id,
-					 lock->un_member.rec_lock.page_id.
-					 fold());
-		ret = ut_fold_ulint_pair(ret, heap_no);
-	} else {
-		/* this check is actually not necessary for continuing
-		correct operation, but something must have gone wrong if
-		it fails. */
-		ut_a(heap_no == 0xFFFF);
-
-		ret = (ulint) lock_get_table(*lock)->id;
-	}
-
-	return(ret);
-#endif
-}
-
-/*******************************************************************//**
-Checks whether i_s_locks_row_t object represents a lock_t object.
-@return TRUE if they match */
-static
-ibool
-locks_row_eq_lock(
-/*==============*/
-	const i_s_locks_row_t*	row,	/*!< in: innodb_locks row */
-	const lock_t*		lock,	/*!< in: lock object */
-	ulint			heap_no)/*!< in: lock's record number
-					or 0xFFFF if the lock
-					is a table lock */
-{
-	ut_ad(i_s_locks_row_validate(row));
-#ifdef TEST_NO_LOCKS_ROW_IS_EVER_EQUAL_TO_LOCK_T
-	return(0);
-#else
-	if (!lock->is_table()) {
-		ut_a(heap_no != 0xFFFF);
-
-		return(row->lock_trx_id == lock->trx->id
-		       && row->lock_page == lock->un_member.rec_lock.page_id
-		       && row->lock_rec == heap_no);
-	} else {
-		/* this check is actually not necessary for continuing
-		correct operation, but something must have gone wrong if
-		it fails. */
-		ut_a(heap_no == 0xFFFF);
-
-		return(row->lock_trx_id == lock->trx->id
-		       && row->lock_table_id == lock_get_table(*lock)->id);
-	}
-#endif
-}
-
-/*******************************************************************//**
-Searches for a row in the innodb_locks cache that has a specified id.
-This happens in O(1) time since a hash table is used. Returns pointer to
-the row or NULL if none is found.
-@return row or NULL */
-static
-i_s_locks_row_t*
-search_innodb_locks(
-/*================*/
-	trx_i_s_cache_t*	cache,	/*!< in: cache */
-	const lock_t*		lock,	/*!< in: lock to search for */
-	uint16_t		heap_no)/*!< in: lock's record number
-					or 0xFFFF if the lock
-					is a table lock */
-{
-	i_s_hash_chain_t*	hash_chain;
-
-	HASH_SEARCH(
-		/* hash_chain->"next" */
-		next,
-		/* the hash table */
-		&cache->locks_hash,
-		/* fold */
-		fold_lock(lock, heap_no),
-		/* the type of the next variable */
-		i_s_hash_chain_t*,
-		/* auxiliary variable */
-		hash_chain,
-		/* assertion on every traversed item */
-		ut_ad(i_s_locks_row_validate(hash_chain->value)),
-		/* this determines if we have found the lock */
-		locks_row_eq_lock(hash_chain->value, lock, heap_no));
-
-	if (hash_chain == NULL) {
-
-		return(NULL);
-	}
-	/* else */
-
-	return(hash_chain->value);
+  ut_ad((heap_no == 0xFFFF) == lock.is_table());
+  if (heap_no == 0xFFFF)
```

Comment:
I would prefer the condition to be `if (lock.is_table())` which should be used uniformly in all places. heap_no is not relevant for table lock and to have the value as "0xFFFF" is good but we better not create functional dependency on it.

Reply:
The `0xFFFF` is special only within this subsystem, returned by the function `wait_lock_get_heap_no()`. We have debug assertions in place (including the one that you quoted) to document this special rule. In non-debug builds, referring to `heap_no == 0xFFFF` should yield better performance than dereferencing `lock`, because that parameter is likely to reside in a register.